pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -0,0 +1,334 @@
+	podsSynced cache.InformerSynced
+
+	// endpointSliceLister is able to list/get pods and is populated by the
+	// shared informer passed to NewController
+	endpointSliceLister discoverylisters.EndpointSliceLister
+	// endpointSlicesSynced returns true if the endpoint slice shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	endpointSlicesSynced cache.InformerSynced
+
+	// nodeLister is able to list/get pods and is populated by the
+	// shared informer passed to NewController
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	nodesSynced cache.InformerSynced
+
+	// triggerTimeTracker is an util used to compute and export the
+	// EndpointsLastChangeTriggerTime annotation.
+	triggerTimeTracker *endpointutil.TriggerTimeTracker
+
+	// Services that need to be updated. A channel is inappropriate here,
+	// because it allowes services with lots of pods to be serviced much
+	// more often than services with few pods; it also would cause a
+	// service that's inserted multiple times to be processed more than
+	// necessary.
+	queue workqueue.RateLimitingInterface
+
+	// maxEndpointsPerSlice references the maximum number of endpoints that
+	// should be added to an EndpointSlice
+	maxEndpointsPerSlice int32
+
+	// workerLoopPeriod is the time between worker runs. The workers
+	// process the queue of service and pod changes
+	workerLoopPeriod time.Duration
+}
+
+// Run will not return until stopCh is closed.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting endpoint controller")
+	defer klog.Infof("Shutting down endpoint controller")
+
+	if !controller.WaitForCacheSync("endpoint_slice", stopCh, c.podsSynced, c.servicesSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	go func() {
+		defer utilruntime.HandleCrash()
+	}()
+
+	<-stopCh
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same service
+// at the same time
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncService(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		klog.V(2).Infof("Error syncing endpoint slices for service %q, retrying. Error: %v", key, err)
+		c.queue.AddRateLimited(key)
```

Comment:
I think utilruntime.HandleError may not record event. So Record Event on the service when error?

Reply:
As far as I can tell, this will only ever get called if syncService fails. I had been triggering an event on the most likely failure there ([reconciler.reconcile()](https://github.com/robscott/kubernetes/blob/endpointslice-controller/pkg/controller/endpointslice/endpointslice_controller.go#L296)), but also added a couple additional events above that should other less likely errors occur (failure to list pods or endpointslices). 

There are still a couple errors that doesn't catch - service fetch or problem parsing the service namespace/name key. Both of those don't have access to a service at that point though so couldn't write an event anywhere meaningful that I can think of. 

If I were to add an event here in `handleErr()` I wouldn't have access to the service either, and to get it I'd have to use the same logic that could potentially result in an error.